src/StreamJsonRpc/Reflection/MessageFormatterRpcMarshaledContextTracker.cs

```diff
@@ -87,6 +87,44 @@ private enum MarshalMode
             MarshallingRealObject = 1,
         }
 
+        internal static (JsonRpcProxyOptions ProxyOptions, JsonRpcTargetOptions TargetOptions)? GetMarshalOptionsForType(Type type)
+        {
+            Requires.NotNull(type, nameof(type));
+
+            foreach ((Type implicitlyMarshaledType, JsonRpcProxyOptions proxyOptions, JsonRpcTargetOptions targetOptions) in ImplicitlyMarshaledTypes)
+            {
+                if (implicitlyMarshaledType == type ||
+                    (implicitlyMarshaledType.IsGenericTypeDefinition &&
+                     type.IsConstructedGenericType &&
+                     implicitlyMarshaledType == type.GetGenericTypeDefinition()))
+                {
+                    return (proxyOptions, targetOptions);
+                }
+            }
+
+            if (type.IsInterface && type.GetCustomAttribute<RpcMarshalableAttribute>() is not null)
```

Comment:
The happy path through this block allocates 2 arrays and potentially many objects. It seems worth caching any Type that passes through the happy path so that each compliant one only pays the penalty once. 
We could also reuse the resulting tuple to avoid a couple more allocations (effectively making this method alloc-free after the first call for a given Type) since although they are mutable, this is an internal method and its callers are not expected to mutate the returned objects.